src/sdk/web_server/http_request.cc

```diff
@@ -0,0 +1,172 @@
+#else
+  char constexpr path_separator = '/';
+  if (result.back() == path_separator) result.resize(result.size() - 1);
+  result.append(path.data(), path.size());
+#endif
+  return result;
+}
+
+std::map<std::string, std::string> parse(const std::string &query) {
+  std::map<std::string, std::string> data;
+  std::regex pattern("([\\w+%]+)=([^&]*)");
+  auto words_begin = std::sregex_iterator(query.begin(), query.end(), pattern);
+  auto words_end = std::sregex_iterator();
+
+  for (std::sregex_iterator i = words_begin; i != words_end; i++) {
+    std::string key = (*i)[1].str();
+    std::string value = (*i)[2].str();
+    data[key] = value;
+  }
+
+  return data;
+}
+}
+
+namespace dsa {
+
+HttpRequest::HttpRequest(
+    WebServer &web_server, boost::asio::ip::tcp::socket socket,
+    http::request<request_body_t, http::basic_fields<alloc_t>> req)
+    : _web_server(web_server),
+      _socket(std::move(socket)),
+      _req(std::move(req)),
+      _resp(std::make_shared<HttpResponse>()) {}
+
+shared_ptr_<HttpResponse> HttpRequest::get_response() { return _resp; }
+
+void HttpRequest::send_bad_response(http::status status,
+                                    std::string const &error) {
+  shared_ptr_<HttpResponse> response = get_response();
+  response->prepare_string_response();
+  response->get_string_response()->result(status);
+  response->get_string_response()->keep_alive(false);
+  response->get_string_response()->set(http::field::server, "dglux_server");
+  response->get_string_response()->set(http::field::content_type, "text/plain");
+  response->get_string_response()->body() = error;
+  response->get_string_response()->prepare_payload();
+  response->prepare_string_serializer();
+  response->string_writer(std::move(_socket));
+}
+
+void HttpRequest::send_file(boost::beast::string_view target) {
+  if (target.empty() || target[0] != '/' ||
+      target.find("..") != std::string::npos) {
+    send_bad_response(http::status::not_found, "File not found\r\n");
+    return;
+  }
+
+  std::string full_path = _doc_root;
+  full_path.append(target.data(), target.size());
+
+  http::file_body::value_type file;
+  boost::beast::error_code ec;
+  file.open(full_path.c_str(), boost::beast::file_mode::read, ec);
+  if (ec) {
+    send_bad_response(http::status::not_found, "File not found\r\n");
+    return;
+  }
+
+  shared_ptr_<HttpResponse> response = get_response();
+  response->prepare_file_response();
+  response->get_file_response()->result(http::status::ok);
+  response->get_file_response()->keep_alive(false);
+  response->get_file_response()->set(http::field::server, "dglux_server");
+  response->get_file_response()->set(http::field::content_type,
+                                     mime_type(full_path));
+  response->get_file_response()->body() = std::move(file);
+  response->get_file_response()->prepare_payload();
+  response->prepare_file_serializer();
+  response->file_writer(std::move(_socket));
```

Comment:
just curious, does this load a large file in memory, and send it? or is it smart enough to only load small buff from the file each time?

Reply:
As far as i understand from the beast document and the usage (passing serializer not the file body itself into async_write as parameter), the latter makes sense to me but i'm gonna verify it anyway.